core/shared/src/main/scala/zio/stm/STM.scala

```diff
@@ -303,10 +299,68 @@ final class STM[+E, +A] private[stm] (
    */
   final def zipWith[E1 >: E, B, C](that: => STM[E1, B])(f: (A, B) => C): STM[E1, C] =
     self flatMap (a => that map (b => f(a, b)))
+
+  private def continueWithM[E1, B](continueM: TExit[E, A] => STM[E1, B]): STM[E1, B] =
+    new STM(
+      (journal, fiberId, stackSize) => {
+        val framesCount = stackSize.incrementAndGet()
+
+        if (framesCount > STM.MaxFrames) {
+          throw new STM.Resumable(self, Stack(continueM))
+        } else {
+          val continued =
+            try {
+              continueM(self.exec(journal, fiberId, stackSize))
+            } catch {
+              case res: STM.Resumable[e, e1, a, b] =>
+                res.ks.push(continueM.asInstanceOf[TExit[e, a] => STM[e1, b]])
+                throw res
+            }
+
+          continued.exec(journal, fiberId, stackSize)
+        }
+      }
+    )
+
+  private def run(journal: STM.internal.Journal, fiberId: Fiber.Id): TExit[E, A] = {
+    type Cont = TExit[Any, Any] => STM[Any, Any]
+
+    val stackSize = new AtomicLong()
+    val stack     = new Stack[Cont]()
+    var current   = self.asInstanceOf[STM[Any, Any]]
+    var result    = null: AnyRef
+
+    while (result eq null) {
+      try {
+        val v = current.exec(journal, fiberId, stackSize)
+
+        if (stack.isEmpty)
+          result = v
+        else {
+          val next = stack.pop()
+          current = next(v)
+        }
+      } catch {
+        case cont: STM.Resumable[_, _, _, _] =>
+          current = cont.stm
+
+          while (!cont.ks.isEmpty) stack.push(cont.ks.pop().asInstanceOf[Cont])
+
+          stackSize.set(0)
+      }
+    }
+
+    result.asInstanceOf[TExit[E, A]]
+  }
 }
 
 object STM {
 
+  private final class Resumable[E, E1, A, B](val stm: STM[E, A], val ks: Stack[internal.TExit[E, A] => STM[E1, B]])
```

Comment:
Should we use https://github.com/scala/scala/blob/2.13.x/src/library/scala/util/control/ControlThrowable.scala for this?

Reply:
I wouldn't go for it. Its representation has been changed in 2.13, from trait inheriting `NoStackTrace` to an abstract class, and according to the docs, it seems that its purpose is also changed a bit. 

Related to this, the reason to avoid `NoStackTrace` in the first place was the possibility to override the suppression via system properties.

Reply:
makes sense 👍